bindings/rust/s2n-tls/src/fingerprint.rs

```diff
@@ -0,0 +1,717 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Methods to fingerprint ClientHellos.
+//!
+//! See [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+
+use crate::{
+    client_hello::ClientHello,
+    error::{Error, Fallible},
+    ffi::*,
+};
+
+use core::ptr::NonNull;
+
+#[non_exhaustive]
+#[derive(Copy, Clone)]
+pub enum FingerprintType {
+    JA3,
+}
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+/// A fingerprint operation.
+pub struct Fingerprint<'a>(&'a mut Builder);
+
+impl Fingerprint<'_> {
+    /// Size of the fingerprint hash.
+    /// See s2n_fingerprint_get_hash_size in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+    pub fn hash_size(&self) -> Result<usize, Error> {
+        self.0.hash_size()
+    }
+
+    /// Calculate the fingerprint hash string.
+    /// See s2n_fingerprint_get_hash in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+    pub fn hash(&mut self) -> Result<&str, Error> {
+        if self.0.hash.is_empty() {
+            let mut output_size = 0;
+            unsafe {
+                s2n_fingerprint_get_hash(
+                    self.0.ptr.as_ptr(),
+                    self.0.hash.capacity() as u32,
+                    self.0.hash.as_mut_ptr(),
+                    &mut output_size,
+                )
+                .into_result()?;
+
+                self.0.hash.as_mut_vec().set_len(output_size as usize);
+            }
+        }
+        Ok(&self.0.hash)
+    }
+
+    /// Size of the raw fingerprint string.
+    ///
+    /// See s2n_fingerprint_get_raw_size in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+    ///
+    /// The size of the raw fingerprint string can't be known without calculating
+    /// the fingerprint for a given ClientHello, so either [Fingerprint::hash()]
+    /// or [Fingerprint::raw()] must be called before this method.
+    pub fn raw_size(&self) -> Result<usize, Error> {
```

Comment:
Just curious: would it be possible to enforce this at compile-time? Like an intermediate Fingerprint that only contains hash() and raw() which returns another Fingerprint that contains raw_size() or something. No idea if this is possible or even desirable though.

Reply:
I considered that, and it might be possible. But I don't think it's desirable. I think it makes the interface more complex for the sake of the less common use case.

Let's call the two structs FingerprintHash and FingerprintFull. The Builder would need to be able to build either FingerprintHash or FingerprintFull depending on whether or not you called set_raw_size. Tricky, but probably doable: I think the easiest solution would be corresponding BuilderHash and BuilderFull builders, and set_raw_size transforms one into the other. But then we also need to transform FingerprintHash into FingerprintFull after `hash` is called and the size of the raw is known. So instead of returning just "&str", `hash` now has to return something like (&str, FingerprintFull). If you only wanted the hash, that FingerprintFull is just discarded. It reminds me of the existing ClientHello::fingerprint_hash and it's awkward str_size return.

Reply:
Oh and I guess that gets even messier, because raw_size would still be unknown even if `set_raw_size` was called. So you'd still get an error despite having a FingerprintFull. But you'd need a FingerprintFull, because you could call `raw` in that state. But `raw` could also fail in that case, if the maximum raw size wasn't set high enough.

Reply:
Yeah that does seem quite complex!